docs/wasm_abi_quirks.md

```diff
@@ -0,0 +1,241 @@
+### Direct vs indirect
+
+There are a couple common ways that aggregates get passed as function parameters at the ABI level in different ABIs.
+
+One is "direct" passing: where every field of the aggregate becomes a single parameter. For example, for the following struct
+
+```rust
+#[repr(C)]
+struct MyStruct {
+    a: u8,
+    // 3 bytes padding
+    b: u32
+}
+
+#[no_mangle]
+extern "C" fn takes_struct(s: MyStruct) {...}
+```
+
+"direct" passing would mean that `takes_struct` is invoked as `wasm.takes_struct(a, b)`.
+
+The LLVM IR to get this result in Wasm looks something like:
+
+```llvm
+; produced by rustc with -Zwasm-c-abi=legacy
+define dso_local void @takes_struct(i8 %x.0, i32 %x.1) unnamed_addr #0 { ... }
+```
+
+This would work through layers of indirection; e.g. a struct with `MyStruct` as its only field would get passed similarly. The idea is that you pick out each scalar value transitively contained in the struct and pass them as arguments one by one.
+
+Another is "indirect" passing, which works similarly to the return value thing where the struct is passed as a pointer. It can be allocated on the heap or the stack, however being able to manipulate the Wasm stack from JS is tricky so it's best to just heap-allocate.
+
+This would end up being invoked something like:
+
+```js
+// Allocate space for the struct with the right size/alignment
+let structAlloc = wasm.diplomat_alloc(8, 4);
+
+ptrWrite(structAlloc, a, 1);
+ptrWrite(structAlloc + 4, b, 4);
+
+wasm.takes_struct(structAlloc);
+
+// Clean up
+wasm.diplomat_free(structAlloc);
+```
+
+The LLVM IR to get this result in Wasm looks something like:
+
+```llvm
+; produced by webassembly-clang:
+%struct.MyStruct = type { i8, i32 }
+define dso_local void @takes_struct(ptr noundef byval(%struct.MyStruct) align 4 %0) #0 { ... }
+
+; produced by rustc with -Zwasm-c-abi=spec
+define dso_local void @takes_struct(ptr byval([8 x i8]) align 4 %x) unnamed_addr #0 { ... }
+```
+
+The [tool conventions] ask for "direct" passing for scalars (including structs that transitively contain a single scalar), and "indirect" for everything else.
+
+### What Rust does
+
+However, Rust doesn't quite do this. Rust _does_ use "direct" passing for scalars and aggregates transitively containing two scalars. However, for aggregates transitively containing more than two scalars, Rust does something that I will call "**padded direct**".
+
+In "padded direct" mode, every transitive scalar field is passed through as an argument just like "direct", however _padding is passed through as well_.
+
+`MyStruct` above contains 3 bytes of padding between the two fields. In "padded direct" mode, this would be invoked as `wasm.takes_struct(a, 0, 0, 0, b)`, with each 0 subbing in for padding. The LLVM IR would look something like:
+
+```llvm
+; Edited from IR produced by rustc with (default) -Zwasm-c-abi=legacy
+%MyStruct = type { i8, [3 x i8], i32 }
+define dso_local void @takes_struct(%MyStruct %0) unnamed_addr #0 { ... }
```

Comment:
Suggestion (optional): consider including the WebAssembly Text (WAT) of the exported function signature, or at least mention how this IR is translated to WebAssembly Text. As a reader of this document, I feel fairly comfortable with WebAssembly Text but not as much in LLVM IR.

For example, one thing that confuses me is that this signature seems to take 5 integers, but 4 of them are `i8`s, and `i8` is not a type in Wasm. Only `i32` and `i64`.

https://webassembly.github.io/spec/core/syntax/types.html

Reply:
The new section on parameter types (previously, the "u64" section) explains this